### ISSUE TITLE
Fix private vol handling during updates (QSB #50)

### ISSUE THREAD

**Issue by tasket** [QubesOS/qubes-issues#5192]
**Qubes OS version**
R4.0

**Affected component(s) or functionality**
dom0 updates i.e. `qubes-dom0-update`

**Brief summary**
The security flaw described in [QSB #50](https://github.com/QubesOS/qubes-secpack/blob/master/QSBs/qsb-050-2019.txt) is that a template reinstall process won't erase the contents of the private volume. Assuming the template was compromised before reinstallation, this might result in compromise after reinstallation via contents in /rw.

**To Reproduce**
`qubes-dom0-update --action=reinstall qubes-template-something-os`

**Expected behavior**
Template's *-private volume is blanked-out, causing the template OS to reinitialize it via /usr/lib/qubes/init/setup-rw.sh on the next template startup.

**Solutions you've tried**
QSB 50 describes a workaround using the more convoluted instructions in the docs (see below).

As for the right way to patch: I don't know of a simple command that can be inserted into `qubes-dom0-update` to achieve a zeroed-out private volume. The `qvm-volume` command doesn't seem to support such a function (and its `resize --force` option appears broken), so the Linux tools `blkdiscard` and `truncate` could be used directly for their respective storage types.

A way to determine storage type would be to parse the output of `qvm-volume ls` to get the storage pool for the target volume, then look up the storage type using `qvm-pool -l`. This provides enough information to use the correct zeroing method (i.e. a volume residing in an 'lvm-thin' pool would be zeroed using `blkdiscard`).

Another way to resolve the issue would be to add a high-level zeroing function in `qvm-volume`. Unfortunately, the storage API doesn't seem to support either zeroing or independent volume removal or a resize that shrinks (in this case, would be shrink to zero size) – please correct this assertion if its not accurate. Adding a zeroing function to the storage layer, however, could be a reasonable approach.

**Relevant [documentation](https://www.qubes-os.org/doc/) you've consulted**
https://www.qubes-os.org/doc/reinstall-template/#manual-method

**Related, [non-duplicate](https://www.qubes-os.org/doc/reporting-bugs/#new-issues-should-not-be-duplicates-of-existing-issues) issues**

**Comment by rustybird** [comment]
> I don't know of a simple command that can be inserted into `qubes-dom0-update` to achieve a zeroed-out private volume. The `qvm-volume` command doesn't seem to support such a function (and its `resize --force` option appears broken) [...]
>
> Unfortunately, the storage API doesn't seem to support either zeroing or independent volume removal or a resize that shrinks (in this case, would be shrink to zero size)

`qvm-volume resize --force vm:volume 0` generally works fine, but there's a redundant safety check [in `lvm_thin`](https://github.com/QubesOS/qubes-core-admin/blob/aa931fdbf73fcd0952a0176873f9e6b174b35491/qubes/storage/lvm.py#L625-L630) and [in `file`](https://github.com/QubesOS/qubes-core-admin/blob/aa931fdbf73fcd0952a0176873f9e6b174b35491/qubes/storage/file.py#L233-L238), so in practice only [`file-reflink` supports downsizing](https://github.com/QubesOS/qubes-core-admin/commit/25e92bd19bc7f3b236a725716b16b08bd1c39608).

We could specify a storage API method `volume.reset()` that preserves the current volume data as a revision (if applicable) and then zeros it. This would also fix #5177.

**Comment by andrewdavidwong** [comment]
CC: @marmarek, @HW42

**Comment by tasket** [comment]
@rustybird Currently, `qvm-volume resize --force vm:private 0` says it won't comply:
```
For your own safety, shrinking of private is disabled (0 < 8799649792). If you really know what you are doing, use `lvresize`
```

Oddly enough, if you omit `--force` it will say the operation could be performed if you include the `--force` option. Assuming this worked at point, maybe restoring its function would be a good route.

Also, `blkdiscard` has recently been added to the lvm.py module to compensate for `lvremove` not generating discards. So the best practice in fixing this issue would be with a patch that's consistent with the `blkdiscard` behavior. I don't think that using a _resize-shrink_ operation by itself would generate discards for freed blocks. (Another way to look at this is that adding discard generation to lvm.py was simple because we didn't have to take shrinking into account as its unsupported.) @brendanhoar

**Comment by hexagonrecursion** [comment]
> Unfortunately, the storage API doesn't seem to support [...] zeroing

I'm not an expert is Qubes OS internals but I think it technically does:
```
f = tempfile.TemporaryFile()
f.truncate(volume.size())
volume.import_data(stream=f)
```
Note: technically whether [truncate](https://docs.python.org/3/library/io.html#io.IOBase.truncate) zero-fills the file contents is platform-dependent.

P.S. I think having some way to shrink a volume to 0 size or completely remove it has some advantages:
1. less IO compared to reading and writing out 2GiB or zeroes.
2. If the user extended the private volume, it would be back to the default size after reinstall (consistent with how the root volume appears to be currently handled).
3. I don't know whether lvm or btrfs are smart enough not to store entire blocks of zeroes (they probably are).

P.P.S. Could we just make the size of 0 special (ie bypass the test for shrinking)? If the user explicitly asks a volume to be shrunk to 0 size, it's pretty clear they don't care about preserving the data inside.

**Comment by rustybird** [comment]
> Currently, `qvm-volume resize --force vm:private 0` says it won't comply:
> 
> ```
> For your own safety, shrinking of private is disabled (0 < 8799649792). If you really know what you are doing, use `lvresize`

That's what I meant by `lvm_thin` having a redundant safety check. This message doesn't come from the general "storage driver agnostic" `qvm-volume` code, but from the specific driver's implementation of `resize()`.

I haven't tested it, but it looks possible to just [remove the LVM specific check](https://github.com/QubesOS/qubes-core-admin/blob/aa931fdbf73fcd0952a0176873f9e6b174b35491/qubes/storage/lvm.py#L625-L630) and [change `lvextend` to `lvresize`](https://github.com/QubesOS/qubes-core-admin/blob/aa931fdbf73fcd0952a0176873f9e6b174b35491/qubes/storage/lvm.py#L758).

> Oddly enough, if you omit `--force` it will say the operation could be performed if you include the `--force` option.

*That* safety check and its error message [comes from general `qvm-volume` code](https://github.com/QubesOS/qubes-core-admin-client/commit/70b15c2eae66acd5896afcbc58fe8108ef0cb0bf#diff-f854b066b3163d1b9ec58150517491b3R202), before any particular driver's `remove()` method is invoked.

> Also, `blkdiscard` has recently been added to the lvm.py module to compensate for `lvremove` not generating discards. So the best practice in fixing this issue would be with a patch that's consistent with the `blkdiscard` behavior. I don't think that using a _resize-shrink_ operation by itself would generate discards for freed blocks.

Oh right, it probably wouldn't.

**Comment by brendanhoar** [comment]
> Also, `blkdiscard` has recently been added to the lvm.py module to compensate for `lvremove` not generating discards. So the best practice in fixing this issue would be with a patch that's consistent with the `blkdiscard` behavior. I don't think that using a _resize-shrink_ operation by itself would generate discards for freed blocks. (Another way to look at this is that adding discard generation to lvm.py was simple because we didn't have to take shrinking into account as its unsupported.) @brendanhoar

Oh yeah. Hmm. 

A "resize" function for lvm-thin LVs isn't as particularly **critical** as it is with file-based volumes.*  That being said, `blkdiscard` can take lists or ranges of logical block numbers to issue discards against. So I suspect the solution (at the volume level) is pretty straightforward, other than being really really sure that the math is correct: blkdiscard [upper-range-of-blocks] then lvresize just under that upper range of blocks. Though I may be assuming how lvresize does things (so perhaps I should look).

OTOH, the filesystem stuff is generally a nightmare, PLUS we have the Qubes context of avoiding the mounting of VM filesystems or even the parsing of VM LV partitions for security model reasons. If one wants something turnkey in dom0, one would probably need to design a process that utilizes disposable "partition and filesystem adjustment" VMs to compact the filesystems/partitions (or even just to verify they are compacted). Alternately...have the user type "ireallyknowwhatiamdoingpleasedestroymyvolume" every time they shrink a LV from dom0. :)

B

* Caveat: I do think that the linux templates and template-based AppVMs should be performing fstrim automatically on shutdown (on only the default r/w volumes for that VM), primarily for maximum storage space recovery; the currently utilized discard mount option doesn't recover as much space due to the LVM cluster-size vs. file-system block-size disparity. I look at the LV volume size primarily as a denial-of-service protection against storage resource exhaustion due to bad planning (and possibly attackers).

**Comment by marmarek** [comment]
> We could specify a storage API method `volume.reset()` that preserves the current volume data as a revision (if applicable) and then resets it. This would also fix #5177.

On Admin API level it's already possible - call `admin.vm.volume.Import+private` with empty or all zeroes input. So, it's just a matter of adding relevant `qvm-volume` sub-command. Also, you can control whether revision should be saved or not with `revisions_to_keep` property.
If for whatever reason (like @noseshimself case) one would like to restore old version, there is `qvm-volume revert`.
